### ISSUE TITLE
Provide consistent sort order in VM lists

### ISSUE THREAD

**Issue by GWeck** [QubesOS/qubes-issues#5678]
**The problem you're addressing (if any)**
While the Qube Manager allows to sort its VM list in several useful ways, at least two VM lists are sorted in fixed order:

- The app menu is sorted showing the VMs in groups determined by their type: DispVMs, Domains (AppVMs and Standalone VMs mixed), Service VMs, and Template VMs. Within each group, the VMs are sorted alphabetically, without case sensitivity.

- In the Copy/Move to other VM operation, the target VM list is sorted quite differently: First, the disposable VMs are listed, and then all other VMs, regardless of their type. Listing is done with case sensitivity, i.e. first all VMs having a name starting with a capital letter are listed alphabetically, and then all VMs having names starting with a lowercase letter.

If there are a large number of VMs, one has to search the intended one, which may be a bit confusing.

**Describe the solution you'd like**
Using always the same sort order would be a step in the right direction. Ideally, this sort order would be the same that is currently used in the Qube manager.

**Where is the value to a user, and who might that user be?**
This would help finding the intended VM in a long list much easier and quicker.

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
#5677

**Comment by hexagonrecursion** [comment]
> Ideally, this sort order would be the same that is currently used in the Qube manager.

Are you suggesting that the sort order everywhere in should change when the user changes the sort order in the Qubes manager?
1. This would be a highly surprising behavior
2. I know I would constantly forget to reset the sort order to the one I'm used to after e.g. sorting by disk usage. I think many others would too.
3. It is possible to start multiple instances of Qubes manager.

I agree that providing a way for the user to affect the sort order in various menus is useful but I think connecting it to the Qubes manager sort order is a bad idea. Ideally there would be a way of manually adjusting the order on a qube-by-qube basis.

**Comment by marmarek** [comment]
I think the main point is using consistent sort order. Not necessary configurable one.

**Comment by GWeck** [comment]
You are right: The advantage of a fixed sort oder is that you will know where to find a certain VM (i.e. "untrussted" is about a third down the list, just like you know that egg rolls are number 8 in your Chinese restaurant's menu).

To be consistent, howver, the Qube Manager should by default display the VMs in just that sort order, and it should be possible to revert to this default if the sort order was changed in the Qube Manager. Using the general sort order in the Qube Manager has the advantage that you see the VMs in this order whenever you look at the Qube Manager and so will quickly learn where to find each VM.

The current sort order in the App menu seems to me to be a good starting point for such a general sort order:

- Grouping the VMs according to their type lets you quickly find the intended one, as you know if you are looking for a template, an AppVM, or so.

- If template-based and standalone VMs are put into the same group or separately into two groups is a secondary matter; there are arguments for both solutions.

- Putting service VMs and template VMs at the end of the list is a good idea, since, for most operations, you will probably use some work domain. The templates for disp VMs should be put at the end, too.

- Having dom0 at the start is also good, since it provides some sort of anchor for the whole list.

- I am not sure whether disp VMs should better be put at the start or at the end of the list. Putting them somewhre in between, however, could probably cause confusion, since their number is bound to change, perhaps even quite rapidly.

The sort order of the VMs in each group could be based on the VM's name, i.e. alphabetically (without case-sensitivity!!!). Sorting first according to colour, as it is now done in the Qube Manager, is also possible, but seems to me a little more counter-intuitive.

One more item: Currently the Q widget shows the running VMs and, addtionally the menu entry "Open Qube Manager", which will be needed to start some other VM. Showing an additional menu entry "Start new VM" would allow to do this without the need to open the Qube Manager. In this case a sorted list of the inactive VMs could be displayed in oder to select one of them for starting.

**Comment by unman** [comment]
On Sat, Feb 22, 2020 at 01:46:09AM -0800, Dr. Gerhard Weck wrote:
> You are right: The advantage of a fixed sort oder is that you will know where to find a certain VM (i.e. "untrussted" is about a third down the list, just like you know that egg rolls are number 8 in your Chinese restaurant's menu).
> 
> To be consistent, howver, the Qube Manager should by default display the VMs in just that sort order, and it should be possible to revert to this default if the sort order was changed in the Qube Manager. Using the general sort order in the Qube Manager has the advantage that you see the VMs in this order whenever you look at the Qube Manager and so will quickly learn where to find each VM.
> 
> The current sort order in the App menu seems to me to be a good starting point for such a general sort order:
> 
> - Grouping the VMs according to their type lets you quickly find the intended one, as you know if you are looking for a template, an AppVM, or so.
> 
> - If template-based and standalone VMs are put into the same group or separately into two groups is a secondary matter; there are arguments for both solutions.
> 
> - Putting service VMs and template VMs at the end of the list is a good idea, since, for most operations, you will probably use some work domain. The templates for disp VMs should be put at the end, too.
> 
> - Having dom0 at the start is also good, since it provides some sort of anchor for the whole list.
> 
> - I am not sure whether disp VMs should better be put at the start or at the end of the list. Putting them somewhre in between, however, could probably cause confusion, since their number is bound to change, perhaps even quite rapidly.
> 
> The sort order of the VMs in each group could be based on the VM's name, i.e. alphabetically (without case-sensitivity!!!). Sorting first according to colour, as it is now done in the Qube Manager, is also possible, but seems to me a little more counter-intuitive.
> 
> One more item: Currently the Q widget shows the running VMs and, addtionally the menu entry "Open Qube Manager", which will be needed to start some other VM. Showing an additional menu entry "Start new VM" would allow to do this without the need to open the Qube Manager. In this case a sorted list of the inactive VMs could be displayed in oder to select one of them for starting.
> 

There is, of course, a real danger of one person's likes setting the
pace for everyone.
None of this affects me, because:
a) I don't use the Menu much (at all)
b) I use KDE, which makes it trivially easy to set the menu as you wish.

Perhaps more effort could go into making the Xfce default as easy to
re-organise as KDE - I believe there are solutions that could be
packaged in default install.

On the specific suggestions:
It makes no sense to treat dispVM Templates the same as Templates,
because users may have many AppVMs set as Templates. I don't know if they
should be with Templates or not.
Putting Service and Templates at the end may be a good idea, but is it
possible to nest them, rather than have a plain list? Again, simple in
KDE.
Users may have a preferred order in the Qube Manager. as I read this
suggestion, this would be *different* from that in the Menu. That is,
they may not see the Menu order in the Manager at all.

Not for the first time, I say this is somewhere we need input from a UX
person.

unman

**Comment by GWeck** [comment]
Clearly, this issue should be addressed by a UX expert. My hope in this respect currently lies with #5679.

As it is now in the App menu, the pre-defined dispVM templates (fedora-30-dvm etc.) are grouped together. I have no idea if they should be grouped the same way as other dispVM templates. Perhaps someone who is using his/her own dispVM templates could comment what would be good?

If a global sort order could be configured in XFCE, this would be a real good solution, as it would not force every user to use the same one, which may be inappropriate. But anyhow, even a sub-optimal general sort order would, in my opinion, be better than having different ones for different operations.

I'll check with KDE, if I get some ideas for XFCE - but I'll do so with caution, as I am no Linux user.

**Comment by GWeck** [comment]
With KDE, the problem is the same: The AppVM, The Move/Copy to VM, and the Qube Manager all use different sort orders - the same ones that are used by XFCE.

**Comment by mfp20** [comment]
From [previous comment](https://github.com/QubesOS/qubes-issues/issues/5679#issuecomment-591059972)

> the only thing I feel the lack of ... is to be able to hide/show some of the VMs (ex: hide/show templates, hide/show red VMs, etc)

A tree-view could be another option.

**Comment by UndeadDevel** [comment]
Related to #8041
